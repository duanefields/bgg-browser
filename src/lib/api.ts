/** Root URL of the BGG Proxy Server */
const BGG_PROXY: string = import.meta.env.VITE_BGG_PROXY

type CollectionResponse = {
  items: {
    item: {
      _objectid: number
      name: {
        _sortindex: string
        __text: string
      }
      yearpublished: string
      image: string
      thumbnail: string
      stats: {
        _minplayers: string
        _maxplayers: string
        _minplaytime: string
        _maxplaytime: string
        _playingtime: string
        _numowned: string
        rating: {
          usersrated: {
            _value: string
          }
          average: {
            _value: string
          }
          bayesaverage: {
            _value: string
          }
          stddev: {
            _value: string
          }
          median: {
            _value: string
          }
          ranks: {
            rank: {
              _type: string
              _id: string
              _name: string
              _friendlyname: string
              _value: string
              _bayesaverage: string
            }[]
            _value: string
          }
        }
      }
      status: {
        _own: string
        _prevowned: string
        _fortrade: string
        _want: string
        _wanttoplay: string
        _wanttobuy: string
        _wishlist: string
        _preordered: string
        _lastmodified: string
      }
      numplays: string
      _objecttype: string
      _subtype: string
      _collid: string
    }[]
  }
  _totalitems: string
  _termsofuse: string
  _pubdate: string
}

type Game = {
  objectId: number
  name: string
  thumbnail: string
  minPlayers: number
  maxPlayers: number
  minPlaytime: number
  maxPlaytime: number
  playtime: number
  yearPublished: number
  averageRating: number
  // rank: number
  numPlays: number
}

const getCollection = async (username: string): Promise<Game[]> => {
  const response = await fetch(
    `${BGG_PROXY}/collection/?username=${username}&stats=1&own=1&version=1&excludesubtype=boardgameexpansion`,
  )

  // Check for errors, the server responds with a 200 OK even if there is an error
  const json = await response.json()
  if (json.errors) {
    throw new Error(json.errors.error.message)
  }

  const data = json as CollectionResponse
  return data.items.item.map((item) => ({
    objectId: Number(item._objectid),
    name: item.name.__text,
    thumbnail: item.thumbnail,
    minPlayers: Number(item.stats._minplayers || 0),
    maxPlayers: Number(item.stats._maxplayers || 0),
    minPlaytime: Number(item.stats._minplaytime || 0),
    maxPlaytime: Number(item.stats._maxplaytime || 0),
    playtime: Number(item.stats._playingtime || 0),
    yearPublished: Number(item.yearpublished),
    averageRating: Number(Number(item.stats.rating.average._value || 0).toFixed(1)),
    // rank: item.stats.rank,
    numPlays: Number(item.numplays),
  }))
}

export { getCollection, BGG_PROXY }
