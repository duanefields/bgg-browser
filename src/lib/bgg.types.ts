export interface RankResponse {
  _type: string
  _id: string
  _name: string
  _friendlyname: string
  _value: string
  _bayesaverage: string
}

export interface ItemResponse {
  _objectid: number
  name: {
    _sortindex: string
    __text: string
  }
  yearpublished: string
  image: string
  thumbnail: string
  stats: {
    _minplayers: string
    _maxplayers?: string
    _minplaytime?: string
    _maxplaytime?: string
    _playingtime?: string
    _numowned: string
    rating: {
      usersrated: {
        _value: string
      }
      average: {
        _value: string
      }
      bayesaverage: {
        _value: string
      }
      stddev: {
        _value: string
      }
      median: {
        _value: string
      }
      ranks: {
        rank: RankResponse[] | { _value: string }
        _value: string
      }
      _value: string
    }
  }
  status: {
    _own: string
    _prevowned: string
    _fortrade: string
    _want: string
    _wanttoplay: string
    _wanttobuy: string
    _wishlist: string
    _preordered: string
    _lastmodified: string
  }
  numplays: string
  _objecttype: string
  _subtype: string
  _collid: string
}

export interface CollectionResponse {
  items: {
    item?: ItemResponse[]
  }
  _totalitems: string
  _termsofuse: string
  _pubdate: string
}

export interface UserResponse {
  user: {
    firstname: { _value: string }
    lastname: { _value: string }
    avatarlink: { _value: string }
    yearregistered: { _value: string }
    lastlogin: { _value: string }
    stateorprovince: { _value: string }
    country: { _value: string }
    webaddress: { _value: string }
    xboxaccount: { _value: string }
    wiiaccount: { _value: string }
    psnaccount: { _value: string }
    battlenetaccount: { _value: string }
    steamaccount: { _value: string }
    traderating: { _value: string }
    _id: string
    _name: string
    _termsofuse: string
  }
}

export interface InvalidUsernameResponse {
  errors: {
    error: {
      message: string
    }
  }
}

export interface CollectionProcessingResponse {
  message: string
}
