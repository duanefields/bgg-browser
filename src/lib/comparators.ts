import { Game } from "../shared.types"

/** sort by name ascending, ignoring leading articles a, an, and the */
export const titleComparator = (a: Game, b: Game) => {
  const aTitle = a.name.replace(/^(a|an|the) /i, "")
  const bTitle = b.name.replace(/^(a|an|the) /i, "")
  return aTitle.localeCompare(bTitle)
}

/** sort by averageRating descending, then by title */
export const ratingComparator = (a: Game, b: Game) => {
  return b.averageRating - a.averageRating || titleComparator(a, b)
}

/** sort by myRating descending, then by title */
export const myRatingComparator = (a: Game, b: Game) => {
  return (b.myRating || 0) - (a.myRating || 0) || titleComparator(a, b)
}

/** sort by numPlays descending, then by title */
export const playsComparator = (a: Game, b: Game) => {
  return b.numPlays - a.numPlays || titleComparator(a, b)
}

/** sort by rank ascending, then by title (though rank should be unique) */
export const rankComparator = (a: Game, b: Game) => {
  return (
    (a.rank || Number.MAX_SAFE_INTEGER) - (b.rank || Number.MAX_SAFE_INTEGER) ||
    titleComparator(a, b)
  )
}
