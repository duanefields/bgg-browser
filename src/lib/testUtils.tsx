import { QueryClient, QueryClientProvider } from "@tanstack/react-query"
import { createRootRoute, createRoute, createRouter, RouterProvider } from "@tanstack/react-router"
import { render } from "@testing-library/react"

/** Renders a collection of children with a Route and QueryClient Provider */
export const renderWithProviders = (children: React.ReactNode) => {
  // wrap the component in a fresh QueryClientProvider to isolate the tests
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  })

  // create a router with a single route that renders the children in a QueryClientProvider
  const rootRoute = createRootRoute()
  const indexRoute = createRoute({
    getParentRoute: () => rootRoute,
    path: "/",
    component: () => <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>,
  })
  const router = createRouter({
    routeTree: rootRoute.addChildren([indexRoute]),
  })

  return render(<RouterProvider router={router} />)
}

/** Renders a collection of children with just a QueryClient Provider */
export const renderWithQueryProvider = (children: React.ReactNode) => {
  // wrap the component in a fresh QueryClientProvider to isolate the tests
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  })

  return render(<QueryClientProvider client={queryClient}>{children}</QueryClientProvider>)
}
