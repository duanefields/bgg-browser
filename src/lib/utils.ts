import { Game } from "../shared.types"

/** Randomly shuffle an array of games */
export const shuffle = (games: Game[]) => {
  return games
    .map((value) => ({ value, sort: Math.random() }))
    .sort((a, b) => a.sort - b.sort)
    .map(({ value }) => value)
}
