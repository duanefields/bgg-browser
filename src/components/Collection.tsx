import { useQuery } from "@tanstack/react-query"
import Fuse from "fuse.js"
import { useMemo } from "react"
import { getCollection } from "../lib/api"
import { Game, PlayerCount, Playtime, SortOrder } from "../shared.types"
import GameCell from "./GameCell"
import {
  titleComparator,
  ratingComparator,
  myRatingComparator,
  playsComparator,
  rankComparator,
  playtimeComparator,
} from "../lib/comparators"
import { shuffle } from "../lib/utils"

interface CollectionProps {
  /** The username of the collection you want to load */
  username: string
  /** The search text to filter the collection */
  searchText?: string
  sort: SortOrder
  players: PlayerCount
  playtime: Playtime
}

const Collection = ({ username, searchText, sort, players, playtime }: CollectionProps) => {
  const query = useQuery({
    queryKey: ["collection", username],
    queryFn: () => getCollection(username),
    retry: (failureCount, error) => {
      console.debug(`Error #${failureCount} fetching collection:`, error.message)
      return error.message === "Collection is being processed, try again later"
    },
  })

  // todo: is this the best way to handle undefined games?
  const emptyGames: Game[] = []
  const games = query.data || emptyGames

  // build the search index
  const index = useMemo(() => {
    console.debug(`Indexing ${games.length} games`)
    console.time("index")
    const fuse = new Fuse<Game>(games, { keys: ["name"], threshold: 0.3 })
    console.timeEnd("index")
    return fuse
  }, [games])

  // filter the list by the fuzzy search text
  let results = games
  if (searchText) {
    results = index.search(searchText).map((result) => result.item)
  }

  // filter the list by the number of players
  if (players > 0) {
    results = results.filter((game) => {
      // if the min or max players is 0, then the game supports any number of players
      if (game.minPlayers === 0 || game.maxPlayers === 0) return true
      // otherwise, check if the player count is within the min/max range
      return players >= game.minPlayers && players <= game.maxPlayers
    })
  }

  // filter by the playtime
  if (playtime > 0) {
    results = results.filter((game) => {
      // if the max playtime is 0 or null, assume the game supports any playtime
      if (game.maxPlaytime === 0 || game.maxPlaytime === null) return true
      // otherwise, check if the playtime is less than or equal to the max playtime
      return playtime >= game.maxPlaytime
    })
  }

  // sort the list by the selected sort order
  console.debug(`Sorting ${results.length} games by ${sort}`)
  console.time("sort")
  if (sort === "random") {
    results = shuffle(results)
  } else {
    results.sort((a, b) => {
      switch (sort) {
        case "name":
          return titleComparator(a, b)
        case "rating":
          return ratingComparator(a, b)
        case "myRating":
          return myRatingComparator(a, b)
        case "plays":
          return playsComparator(a, b)
        case "rank":
          return rankComparator(a, b)
        case "playtime":
          return playtimeComparator(a, b)
      }
    })
  }
  console.timeEnd("sort")

  return (
    <div>
      <h2>{username}'s Collection</h2>
      {query.isLoading && <div>Loading...</div>}
      {query.isError && <div>Error: {query.error.message}</div>}
      {query.isSuccess && (
        <div>
          <div>
            Showing {results.length.toLocaleString()} of {games.length.toLocaleString()} games
          </div>
          {results?.map((game) => (
            <div key={game.collectionId}>
              <GameCell game={game} />
            </div>
          ))}
        </div>
      )}
    </div>
  )
}

export default Collection
