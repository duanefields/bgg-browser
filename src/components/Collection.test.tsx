import { QueryClient, QueryClientProvider } from "@tanstack/react-query"
import { render, screen } from "@testing-library/react"
import { http, HttpResponse } from "msw"
import { setupServer } from "msw/node"
import { afterAll, afterEach, beforeAll, expect, it } from "vitest"
import { BGG_PROXY } from "../lib/api"
import invalidUsername from "../test/invalidUsername.json"
import pandyandy from "../test/pandyandy.json"

import Collection from "./Collection"

const server = setupServer(
  http.get(`${BGG_PROXY}/collection`, ({ request }) => {
    const url = new URL(request.url)
    const username = url.searchParams.get("username")
    switch (username) {
      case "pandyandy":
        return HttpResponse.json(pandyandy)
      default:
        return HttpResponse.json(invalidUsername)
    }
  }),
)

beforeAll(() => server.listen())
afterEach(() => server.resetHandlers())
afterAll(() => server.close())

const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      retry: false,
    },
  },
})

const renderWithQueryProvider = (children: React.ReactNode) => {
  return render(<QueryClientProvider client={queryClient}>{children}</QueryClientProvider>)
}

it("Should render a loading state", async () => {
  renderWithQueryProvider(<Collection username="pandyandy" />)
  expect(screen.getByText("Loading...")).toBeVisible()
})

it("Should render an error state", async () => {
  renderWithQueryProvider(<Collection username="nonexistent" />)
  expect(await screen.findByText("Error: Invalid username specified")).toBeVisible()
})

it("Should render the collection", async () => {
  renderWithQueryProvider(<Collection username="pandyandy" />)
  expect(await screen.findByText("Forbidden Island")).toBeVisible()
})

it("Should match snapshot", async () => {
  const collection = renderWithQueryProvider(<Collection username="pandyandy" />)
  await screen.findByText("Forbidden Island")
  expect(collection).toMatchSnapshot()
})
