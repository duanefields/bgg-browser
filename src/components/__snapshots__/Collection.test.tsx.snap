// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <h2>
          pandyandy
          's Collection
        </h2>
        <div>
          <div>
            <h3>
              Dixit
            </h3>
          </div>
          <div>
            <h3>
              Dixit: Journey
            </h3>
          </div>
          <div>
            <h3>
              Forbidden Island
            </h3>
          </div>
          <div>
            <h3>
              Pandemic
            </h3>
          </div>
          <div>
            <h3>
              The Resistance
            </h3>
          </div>
          <div>
            <h3>
              Skull
            </h3>
          </div>
          <div>
            <h3>
              Tsuro of the Seas
            </h3>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <h2>
        pandyandy
        's Collection
      </h2>
      <div>
        <div>
          <h3>
            Dixit
          </h3>
        </div>
        <div>
          <h3>
            Dixit: Journey
          </h3>
        </div>
        <div>
          <h3>
            Forbidden Island
          </h3>
        </div>
        <div>
          <h3>
            Pandemic
          </h3>
        </div>
        <div>
          <h3>
            The Resistance
          </h3>
        </div>
        <div>
          <h3>
            Skull
          </h3>
        </div>
        <div>
          <h3>
            Tsuro of the Seas
          </h3>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
